src/class/impl_.rs

```diff
@@ -0,0 +1,270 @@
+// Copyright (c) 2017-present PyO3 Project and Contributors
+
+use crate::{derive_utils::PyBaseTypeUtils, ffi, PyMethodDefType, PyNativeType};
+use std::{marker::PhantomData, thread};
+
+/// This type is used as a "dummy" type on which dtolnay specializations are
+/// applied to apply implementations from #[pymethods] & #[pyproto]
+pub struct PyClassImplCollector<T>(PhantomData<T>);
+
+impl<T> PyClassImplCollector<T> {
+    pub fn new() -> Self {
+        Self(PhantomData)
+    }
+}
+
+impl<T> Default for PyClassImplCollector<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T> Clone for PyClassImplCollector<T> {
+    fn clone(&self) -> Self {
+        Self::new()
+    }
+}
+
+impl<T> Copy for PyClassImplCollector<T> {}
+
+/// Implements the underlying functionality of `#[pyclass]`, assembled by various proc macros.
+///
+/// Users are discouraged from implementing this trait manually; it is a PyO3 implementation detail
+/// and may be changed at any time.
+pub trait PyClassImpl: Sized {
+    /// This handles following two situations:
+    /// 1. In case `T` is `Send`, stub `ThreadChecker` is used and does nothing.
+    ///    This implementation is used by default. Compile fails if `T: !Send`.
+    /// 2. In case `T` is `!Send`, `ThreadChecker` panics when `T` is accessed by another thread.
+    ///    This implementation is used when `#[pyclass(unsendable)]` is given.
+    ///    Panicking makes it safe to expose `T: !Send` to the Python interpreter, where all objects
+    ///    can be accessed by multiple threads by `threading` module.
+    type ThreadChecker: PyClassThreadChecker<Self>;
+
+    fn for_each_method_def(_visitor: impl FnMut(&PyMethodDefType)) {}
+    fn get_new() -> ffi::newfunc {
```

Comment:
How about making the return type `Option<ffi::newfunc>` and use `unwrap_or(fallback_new)` in the caller?
Just a minor detail, though.

Reply:
Agreed, I think that's more consistent with `get_call` in that way.